Asynchronous harvesters

marktriggs opened the pull request:
This PR moves the three threaded harvesters (IA, UVA, NYPL) from KriKri into Heidrun.

I brought the original commits across verbatim in commit a7ac3a8 then applied James's other changes (to rename Fedora to UVA and pull down IA files records) over the top.  Finally, there are two commits from me to fix the module names and and get the specs running again.

anarchivist commented:
For the record, this will fail until dpla/krikri#208 is merged. I've created a temporary branch of Heidrun (`test-async-harvesters`) based on this pull request that's pulling in Krikri from HM's `asynchronous-harvesters` branch, and tests are running here: https://travis-ci.org/dpla/heidrun/builds/98571558

anarchivist commented:
Hmm, looks like it's failing anyhow, because of something related to Webmock... https://travis-ci.org/dpla/heidrun/builds/98572241

marktriggs commented:
Ah, apologies.  There's a small hostname difference between running rspec manually & when it runs from CI that was causing my Webmock rule to miss.

I've updated this PR, and sent a new one for that test branch just in case that makes it easier:

https://github.com/dpla/heidrun/pull/39

anarchivist commented:
Ah, great. That actually looks like it's adequate to make this pass. I'm :+1: from where I stand, and I'll pass to @markbreedlove for another look with that in mind (although we're quickly running out of hours in the day!)

anarchivist commented:
Also, I'm in favor of *not* squashing these commits.

markbreedlove commented on app/harvesters/ia_harvester.rb:
```diff
@@ -0,0 +1,178 @@
+
+    # wait for the requests to complete so we don't hit the server
+    # harder than intended
+    batch.each { |r| r[:meta_request].join }
```
I made a comment in https://github.com/dpla/KriKri/pull/208 referencing this line. I think that we'll have an exception bubble up here when `#join` is called if there was an exception thrown by one of the Faraday calls in `AsyncUriGetter::Request#http`.

markbreedlove commented:
Looks great, with my only question being one of exception handling, noted above.

markbreedlove commented:
(See further discussion in https://github.com/dpla/KriKri/pull/208)...

We've discussed the responsibility of the harvester to handle exceptions that bubble up from the Thread in the AsyncUriGetter.  There are already some tests included here for bad requests, so as long as we can take a careful look at what's here and ensure that connection errors like bad network sockets are handled properly, then this should be good to go.

no-reply commented:
Testing the `UVAHarvester`:

  - `#record_ids` is unimplemented (`NotImplementedError`). 
    - If there's not an efficient solution for getting record ids, I think `records.map { |rec| rec.id }` (or similar) is an acceptable implementation.
  - I have some documentation improvements, mainly to clarify what URI to pass; I need push access to the `hudmol` repository to add my commit to this PR; otherwise, I can close this and submit the revised PR. (cc: @jambun).
  - Should this harvester be able to accept a set of collections to harvest as part of the same activity? If I understand correctly, we would need to run three separate harvests from a single fedora endpoint, as is.

no-reply commented:
A UVA testing note that hopefully does not require changes in the PR:

I'm seeing a bunch of `500` status code responses when harvesting UVA. For the most part, records seem to work, but a non-trivial amount (at least a few dozen out of 5500), failed with:

```
Couldn't get record uva-lib_333364-mods
	500 -- javax.ws.rs.WebApplicationException: org.fcrepo.server.errors.HttpServiceNotFoundException: [DefaultExternalContentManager] returned an error.  The underlying error was a org.fcrepo.server.errors.GeneralException  The message was  "Error getting http://transform.lib.virginia.edu:8084/saxon/SaxonServlet?source=http%3A%2F%2Ffedora-prod02.lib.virginia.edu%3A8080%2Ffedora%2Fget%2Fuva-lib%3A333364%2FdescMetadata%2F2013-06-12T19%3A12%3A42.717Z&style=http://fedora-prod02.lib.virginia.edu:8080/fedora/get/uva-lib:modsSDep/XSLT&pid=uva-lib%3A333364&clear-stylesheet-cache=yes"  .  
	at org.fcrepo.server.rest.BaseRestResource.handleException(BaseRestResource.java:168)
	at org.fcrepo.server.rest.MethodResource.invokeSDefMethodUsingGET(MethodResource.java:104)
	at sun.reflect.GeneratedMethodAccessor68.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:175)
	at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:67)
	at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:163)
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:111)
	at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:71)
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:111)
	at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:63)
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:689)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:647)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:638)
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:309)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:425)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:590)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:290)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.fcrepo.server.security.servletfilters.FilterRestApiFlash.doFilter(FilterRestApiFlash.java:79)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:233)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:191)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:563)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:127)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:109)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:293)
	at org.apache.coyote.ajp.AjpAprProcessor.process(AjpAprProcessor.java:448)
	at org.apache.coyote.ajp.AjpAprProtocol$AjpConnectionHandler.process(AjpAprProtocol.java:403)
	at org.apache.tomcat.util.net.AprEndpoint$SocketProcessor.run(AprEndpoint.java:2213)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
	at java.lang.Thread.run(Thread.java:619)
Caused by: org.fcrepo.server.errors.HttpServiceNotFoundException: [DefaultExternalContentManager] returned an error.  The underlying error was a org.fcrepo.server.errors.GeneralException  The message was  "Error getting http://transform.lib.virginia.edu:8084/saxon/SaxonServlet?source=http%3A%2F%2Ffedora-prod02.lib.virginia.edu%3A8080%2Ffedora%2Fget%2Fuva-lib%3A333364%2FdescMetadata%2F2013-06-12T19%3A12%3A42.717Z&style=http://fedora-prod02.lib.virginia.edu:8080/fedora/get/uva-lib:modsSDep/XSLT&pid=uva-lib%3A333364&clear-stylesheet-cache=yes"  .  
	at org.fcrepo.server.storage.DefaultExternalContentManager.getExternalContent(DefaultExternalContentManager.java:155)
	at org.fcrepo.server.access.dissemination.DisseminationService.assembleDissemination(DisseminationService.java:684)
	at org.fcrepo.server.access.DefaultAccess.getDissemination(DefaultAccess.java:475)
	at org.fcrepo.server.rest.MethodResource.invokeSDefMethodUsingGET(MethodResource.java:95)
	... 47 more
Caused by: org.fcrepo.server.errors.GeneralException: Error getting http://transform.lib.virginia.edu:8084/saxon/SaxonServlet?source=http%3A%2F%2Ffedora-prod02.lib.virginia.edu%3A8080%2Ffedora%2Fget%2Fuva-lib%3A333364%2FdescMetadata%2F2013-06-12T19%3A12%3A42.717Z&style=http://fedora-prod02.lib.virginia.edu:8080/fedora/get/uva-lib:modsSDep/XSLT&pid=uva-lib%3A333364&clear-stylesheet-cache=yes
	at org.fcrepo.server.storage.DefaultExternalContentManager.get(DefaultExternalContentManager.java:183)
	at org.fcrepo.server.storage.DefaultExternalContentManager.getFromWeb(DefaultExternalContentManager.java:332)
	at org.fcrepo.server.storage.DefaultExternalContentManager.getExternalContent(DefaultExternalContentManager.java:149)
	... 50 more
Caused by: java.io.IOException: Request failed [500 Internal Server Error]
	at org.fcrepo.common.http.WebClient.get(WebClient.java:211)
	at org.fcrepo.common.http.WebClient.get(WebClient.java:136)
	at org.fcrepo.server.storage.DefaultExternalContentManager.get(DefaultExternalContentManager.java:172)
	... 52 more
```

Anyone have ideas about what is going on here? @guegueng we may need to communicate with UVA about these errors.

markbreedlove commented:
Sorry not to have asked this earlier, but is there any provision for picking out specific Internet ARchive "collections?"  See `collections` in the ingestion profile from our legacy system: https://github.com/dpla/ingestion/blob/v33.8.2/profiles/ia.pjs ... Can I do something similar with this new harvester?

anarchivist commented:
@markbreedlove Yep, you can pass them as URI parameters themselves as part of the value for the `:uri` key:

```ruby
[13] pry(main)> opts = {uri: "http://archive.org/advancedsearch.php?q=collection%3A%28blc%20OR%20cambridgepubliclibrary%20OR%20clemson%20OR%20bostonpubliclibrary%20OR%20regionaldigitizationmass%20OR%20getty%20OR%20medicalheritagelibrary%20OR%20MontanaStateLibrary%20OR%20yivoinstitutelibrary%20OR%20guggenheimlibrary%20OR%20lbiperiodicals%20OR%20frickartreferencelibrary%29&output=json&rows=1000&fl%5B%5D=identifier"}
=> {:uri=>
  "http://archive.org/advancedsearch.php?q=collection%3A%28blc%20OR%20cambridgepubliclibrary%20OR%20clemson%20OR%20bostonpubliclibrary%20OR%20regionaldigitizationmass%20OR%20getty%20OR%20medicalheritagelibrary%20OR%20MontanaStateLibrary%20OR%20yivoinstitutelibrary%20OR%20guggenheimlibrary%20OR%20lbiperiodicals%20OR%20frickartreferencelibrary%29&output=json&rows=1000&fl%5B%5D=identifier"}
[14] pry(main)> IaHarvester.enqueue(opts)
created activity 84
enqueued to harvest
```

anarchivist commented on app/harvesters/ia_harvester.rb:
```diff
@@ -0,0 +1,178 @@
+  #
+  #   - threads:     The number of records to fetch asynchronously
+  #                  in a batch (default: 10)
+  #   - name:        See Krikri::Harvester#initialize.  Defaults to "ia"
```
Looks like this default is not actually getting set, unless i'm missing something.

anarchivist replied:
It should be set to `"internet_archive"`, instead of `"ia"`

anarchivist commented on app/harvesters/uva_harvester.rb:
```diff
@@ -0,0 +1,153 @@
+  #
+  #   - threads:     The number of records to fetch asynchronously
+  #                  in a batch (default: 10)
+  #   - name:        See Krikri::Harvester#initialize.  Defaults to "uva"
```
Looks like this default is not actually getting set, unless I'm missing something.

anarchivist replied:
This should be set to `"virginia"`, not `"uva"`

anarchivist commented:
For the UVA harvester, it should be possible to have the harvester take multiple collection PIDs as an parameter rather than requiring separate invocations for each collection.

anarchivist commented on app/harvesters/ia_harvester.rb:
```diff
@@ -0,0 +1,178 @@
+  # @example
+  #    Typical instantiation, good for most cases:
+  #
+  #      IaHarvester.new(uri: 'http://archive.org/...')
```
Should the documentation specify which URI parameters should not be included (e.g. `rows`, since this is set by `:threads`)?

markbreedlove replied:
Also related to the `uri` parameter, we harvest specific collections from Internet Archive, which we specify with a querystring parameter as follows:
```
?q=collection%3A%28blc%20OR%20cambridgepubliclibrary%20OR%20clemson%20OR%20bostonpubliclibrary%20OR%20regionaldigitizationmass%20OR%20getty%20OR%20medicalheritagelibrary%20OR%20MontanaStateLibrary%20OR%20yivoinstitutelibrary%20OR%20guggenheimlibrary%20OR%20lbiperiodicals%20OR%20frickartreferencelibrary%29
```
... which decodes to:
```
collection:(blc OR cambridgepubliclibrary OR clemson OR bostonpubliclibrary OR regionaldigitizationmass OR getty OR medicalheritagelibrary OR MontanaStateLibrary OR yivoinstitutelibrary OR guggenheimlibrary OR lbiperiodicals OR frickartreferencelibrary)
```

As it is now, we have to construct the URI and paste it into the `uri`, including that long string for the collections. It would be ideal to have this constructed for us if we pass in a `collections` array parameter.

anarchivist commented on app/harvesters/nypl_harvester.rb:
```diff
@@ -0,0 +1,238 @@
+  # - uri:        See Krikri::Harvester#initialize.
+  #               Defaults to "http://api.repo.nypl.org/api/v1'"
+  # - name:       See Krikri::Harvester#initialize.  Defaults to "nypl"
+  # - batchsize:  The number of records to fetch with each API request.
```
Should this specify which NYPL API parameter it matches to up here?

markbreedlove commented on app/harvesters/nypl_harvester.rb:
```diff
@@ -0,0 +1,238 @@
+  #               Defaults to 10.
+  # - threads:    The number of record pages to fetch in parallel.
+  #               Defaults to 5.
+  #
```
It would be good if the `apikey` parameter referenced above in the example had a line somewhere around here.

anarchivist commented on app/harvesters/ia_harvester.rb:
```diff
@@ -0,0 +1,200 @@
+      opts: {
+        collections: { type: :string, multiple_ok: true, required: true },
+        threads: { type: :integer, required: false },
+        name: { type: :string, required: false },
```
`:name` should not be included in `#expected_opts` (see [here](https://github.com/dpla/KriKri/blob/develop/lib/krikri/harvester.rb#L150-L181))

anarchivist replied:
Unless... `:name` is actually `opts[:ia][:name]`...?

anarchivist commented on app/harvesters/nypl_harvester.rb:
```diff
@@ -0,0 +1,249 @@
+  #
+  # Parameters to 'opts':
+  #
+  # - uri:        See Krikri::Harvester#initialize.
```
`:uri` and `:name` should not be included in `#expected_opts` (see [here](https://github.com/dpla/KriKri/blob/develop/lib/krikri/harvester.rb#L150-L181))

anarchivist commented on app/harvesters/uva_harvester.rb:
```diff
@@ -0,0 +1,157 @@
+      key: :uva,
+      opts: {
+        threads:  { type: :integer, required: false },
+        name:  { type: :string, required: false },
```
`:name` should not be included in `#expected_opts` (see [here](https://github.com/dpla/KriKri/blob/develop/lib/krikri/harvester.rb#L150-L181))

anarchivist commented:
I'd like to clarify how `:name` is set on the three harvesters in the PR - is it expected to be in `opts[:name]` or `opts[:harvester][:name]`?

anarchivist commented:
@no-reply @markbreedlove Re [discussion above](https://github.com/dpla/heidrun/pull/38#commitcomment-16359673), do we need a new release of Krikri to test?

markbreedlove commented:
@anarchivist: re. the Krikri release, @no-reply is correct that it's not necessary, because we can test with a `heidrun` branch that alters the Gemfile to pull Krikri from the `develop` GitHub branch.

anarchivist commented:
@jambun I think the one thing we're still missing on this is [this](https://github.com/dpla/heidrun/pull/38#issuecomment-187439675):

> For the UVA harvester, it should be possible to have the harvester take multiple collection PIDs as an parameter rather than requiring separate invocations for each collection.

jambun commented:
@anarchivist I've pushed a couple of commits for the collection PIDs change (with a rebase).

I hope it looks ok.

anarchivist commented:
@jambun Looks like the tests are failing on Travis: https://travis-ci.org/dpla/heidrun/builds/114137170

jambun commented:
Sorry for the false laziness @anarchivist, fixups pushed, fingers crossed.

anarchivist commented:
I discovered in testing the IA harvester that the `start` parameter for the offset won't actually paginate the requests, which leads to harvesting the same `:threads` records over and over. We'll need to switch this to using the `page` parameter, which is 1-indexed (when `rows=10`, page 1 contains items 1-10 in the response). Per Slack discussion, I'm passing this back over to @marktriggs.

marktriggs commented:
Right!  I've:

  * Switched it to use the `page` parameter and updated the tests

  * Added the extra debugging lines from @anarchivist 

  * Rebased with the `develop` branch to pick up the changes to rspec & rake versions and the webmock config changes

Unit tests are passing and hopefully Travis will smile upon me :)

anarchivist commented on app/harvesters/uva_harvester.rb:
```diff
@@ -0,0 +1,237 @@
+  include Krikri::Harvester
+
+  DEFAULT_URI_TEMPLATE =
+    'http://fedoraproxy.lib.virginia.edu/fedora/objects/uva-lib%3A[PID]/methods/uva-lib%3AmetsSDef/getMETS'
```
Can this be changed so the template doesn't expect `uva-lib:` as a prefix, so we can pass PIDs like `uva-lib:2190487` directly to it?

anarchivist commented:
OK, in testing the UVA harvester again, I found another set of changes that need to be made based on how it handles item record identifiers. It's a little convoluted other than the minor URI template change I suggested which is why I'm not commenting on the diff (as I'm not entirely certain how these changes would flow together)

To get the item records, we're looking at the `<mets:dmdSec/>` element, and grab both the `ID` attribute and the `<mets:mdRef/>`'s  `xlink:href` attribute. If I'm interpreting the behavior of the `UVAHarvester` methods correctly, the URI for the created `Krikri::OriginalRecord` is based on the former (`//mets:dmdSec[@ID]`), which provides values like `uva-lib_332086-mods`.

However, in the case of UVA, the actual *identifier* of the items in their Fedora 3 instance is in the PID format, like `uva-lib:332086`. This is important to change, since the URIs we mint need to correspond with those minted the old system.

The other minor note is that we can successfully dereference MODS records for individual items by knowing their PID, without having to having to parse the entire METS for one or more collections, using a URI template like the following: `http://fedoraproxy.lib.virginia.edu/fedora/objects/[PID]/methods/uva-lib%3AmodsSDef/getMODS`

anarchivist commented:
Hm, also, re: IA, once it gets to the 1000th page of results it seems to fail (e.g. [this URI](http://archive.org/advancedsearch.php?fl%5B%5D=identifier&output=json&q=collection:(cambridgepubliclibrary%20OR%20clemson%20OR%20bostonpubliclibrary%20OR%20regionaldigitizationmass%20OR%20getty%20OR%20medicalheritagelibrary%20OR%20MontanaStateLibrary%20OR%20yivoinstitutelibrary%20OR%20guggenheimlibrary%20OR%20lbiperiodicals%20OR%20frickartreferencelibrary)&page=1001&rows=10)). This is the message that's returned:

```
"RANGE_OUT_OF_BOUNDS: paging is only supported through 10000 results; scraping is supported through the Scraping API, see https://archive.org/help/aboutsearch.htm  or, you may request up to 100000000 results at one time if you do NOT specify any page. For best results, do NOT specify sort (sort may be automatically disabled for very large queries)."
```

marktriggs commented:
I've reworked the IA harvester to use the new "scrape" API, and have modified UVA to use the same IDs as the old system.

anarchivist commented:
@marktriggs see above

marktriggs commented:
Can't rely on *anything* these days :)
... except regular expressions!

I've just patched the UVA harvester to pull the ID from the URL.

anarchivist commented on app/harvesters/uva_harvester.rb:
```diff
@@ -0,0 +1,248 @@
+      batch.flat_map do |record|
+        record[:request].with_response do |response|
+          unless response.status == 200
+            msg = "Couldn't get record #{record[:id]}"
```
This actually isn't reporting the record ID that fails - I'm seeing a log message like the following: `E, [2016-03-15T23:10:10.503649 #3766] ERROR -- : [2016-03-15 23:10:10 -0400] [3766] [#<Krikri::Logger:0x007fa34c96f658
>] Couldn't get record`

anarchivist commented:
In the UVA harvester, the default value for the `:name` parameter isn't propagating properly to the ID minter:

```
[6] pry(main)> Krikri::Activity.find(140).entity_uris.force.last
  Krikri::Activity Load (0.9ms)  SELECT  "krikri_activities".* FROM "krikri_activities"  WHERE "krikri_activities"."id" = $1 LIMIT 1  [["id", 140]]
=> "http://ldp.staging.dp.la/ldp/original_record/6bb5d00e98fe3b73da56f1720a289653"
[7] pry(main)> Digest::MD5.hexdigest("uva-lib:328435")
=> "6bb5d00e98fe3b73da56f1720a289653"
[8] pry(main)> Digest::MD5.hexdigest("virginia--uva-lib:328435")
=> "3f4722984f02d0b41b680e792add20cf"
```

marktriggs commented:
@anarchivist I've addressed both of those issues now, thanks.  I hadn't realized that the harvest name was used in minting the IDs, but that makes sense.

Regarding the log message, I haven't done a full harvest so I haven't actually seen that fire in the wild yet, and it's so simple that I've almost certainly messed it up.  So don't hesitate to let me know if it still doesn't look right :)

anarchivist commented:
Oy... I spoke to soon about how to fetch the identifiers for UVA, upon further review. I'm *really* sorry about this. :disappointed:  Basically, we need to introduce some logic to handle conditionally handle the presence of specific elements. This behavior reflects [how the logic was implemented in the old system](https://github.com/dpla/ingestion/blob/develop/lib/fetchers/uva_fetcher.py#L16-L30).  The logic is as follows:

* If `/mods:identifier[@type='uri']` exists, take that value
* If it doesn't exist, take the value from `/mods:location/mods:url[@usage='primary display']`.

marktriggs commented:
@anarchivist No worries!  I think I've captured it now.

Am I right to think that the identifier will end up being a URL when we have to go to the location URL for the identifier?  I'm guessing your brief freakout on Slack was when you noticed this but then realized that the old system was already using URLs in that  case too?

anarchivist commented:
> Am I right to think that the identifier will end up being a URL when we have to go to the location URL for the identifier? I'm guessing your brief freakout on Slack was when you noticed this but then realized that the old system was already using URLs in that case too?

Yes, the identifier will be a URL when we use `/mods:location/mods:url[@usage='primary display']`. Oddly, `/mods:identifier[@type='uri']` seems to *not* be a URI, but rather a bare Fedora PID like `uva-lib:nnnnnn`.

anarchivist commented:
@marktriggs I'm signing off on UVA. :white_check_mark: But IA needs a minor fix!

IA harvests are failing when there's a connection timeout:

```
E, [2016-03-16T21:37:51.724634 #15750] ERROR -- : [2016-03-16 21:37:51 -0400] [15750] [#<Krikri::Logger:0x007f9f17a306c0>] Error performing Activity: 148
Connection timed out - connect(2) for "ia600508.us.archive.org" port 80
["/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:879:in `initialize'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:879:in `open'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:879:in `block in connect'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/timeout.rb:76:in `timeout'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:878:in `connect'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:863:in `do_start'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:852:in `start'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:1369:in `request'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/2.1.0/net/http.rb:1128:in `get'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:80:in `perform_request'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:40:in `block in call'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:87:in `with_net_http_connection'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/adapter/net_http.rb:32:in `call'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response/follow_redirects.rb:76:in `perform_with_redirection'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response/follow_redirects.rb:64:in `call'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/request/retry.rb:116:in `call'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:139:in `build_response'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/connection.rb:377:in `run_request'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/connection.rb:140:in `get'", "/home/dpla/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/krikri-0.12.0/lib/krikri/async_uri_getter.rb:97:in `block in start_request'"]
```

I think this is happening once there's a `*_meta.xml* in hand, and I'm guessing from the stacktrace that this is happening in `Krikri::AsyncUriGetter`. Could we log these failures and allow the harvester to continue, rather than forcing the entire job to fail?

anarchivist commented:
Part of the motivation there is that some of IA's petabox servers lose connectivity for a bit, and we don't want a single item failure to throw off the harvest. Retries would be nice but that's probably a larger pocket of work.

marktriggs commented:
Cool, that makes sense, yeah.  I'll assign this back to me for the IA changes.

marktriggs commented:
@anarchivist I've submitted a PR against KriKri for some changes to support the new IA behavior.

marktriggs commented:
... and I've switched the IA harvester to use the new feature now.

anarchivist commented:
![](https://media.giphy.com/media/xT77XRp55yPwHkd5Nm/giphy.gif)
